### PR TITLE
refactor: [M3-8623] - Introduce TanStack Router

### PR DESCRIPTION
**Pull request by abailly-akamai**
Screenshot 2024-10-03 at 11 45 08

## Description 📝
This (large) PR introduces `TanStack Router` to Cloud Manager. 

https://tanstack.com/router/latest

It prepares a full migration from our old, outdated v5 `react-router-dom`  to the new Routing API.

The benefits of this migration are multiple:
- very low dependency
- 100% inferred TypeScript support
- Typesafe navigation
- Nested Routing and layout routes
- Built-in Route Loaders w/ SWR Caching
- Designed for client-side data caches (TanStack Query, SWR, etc.)
- Automatic route prefetching
- Asynchronous route elements and error boundaries
- File-based Route Generation
- Typesafe JSON-first Search Params state management APIs
- Path and Search Parameter Schema Validation
- Search Param Navigation APIs
- Custom Search Param parser/serializer support
- Search param middleware
- Route matching/loading middleware

> [!IMPORTANT]
> This initial PR does not (read **_should not_**) affect the current Cloud Manager routing experience. It only aims to achieve the following:
- consistent file splitting/lazy loading 
- centralized routes (parity with existing routes

I do not expect to have 100% route parity with this PR, I however did my very best get as close as possible. Subsequent PRs will rollout the actual new routing by updating legacy utils with the TanStack ones (params, location, history, etc)

## Changes  🔄
- Create lazy exports for routed components
- Create all existing routes in centralized location
- Configure root and routing architecture

### Prerequisites
- pull code locally and run `yarn` to get new dependency

### Verification steps
- Confirm this PR does not introduce any functional change or regression
- Switch the routing in `<App />` to very basic functionality
  - **important** because we only defined the routes and have not replaced the utils, the experience may be very buggy for many routes - using MSW legacy handlers will help a bit with that matter

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/Root.styles.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    '&.mlMain': {
+      [theme.breakpoints.up('lg')]: {
+        maxWidth: '78.8%',
+      },
+    },
+    flex: 1,
+    maxWidth: '100%',
+    position: 'relative',
+  },
+}));
```
Comment (abailly-akamai) [inline packages/manager/src/Root.styles.ts:85]: All existing styles from `<MainContent />` extracted to their own file - this also isn't affecting the current experience

File: packages/manager/src/Root.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                </main>
+              </div>
+            </NotificationProvider>
+            <Footer desktopMenuIsOpen={desktopMenuIsOpen} />
+          </ComplianceUpdateProvider>
+        </SwitchAccountSessionProvider>
+      </SessionExpirationProvider>
+    </div>
+  );
+};
```
Comment (abailly-akamai) [inline packages/manager/src/Root.tsx:130]: I decoupled the root from the <MainContent /> so the saparation of concern between routing and composition is more clear. This is **not** affecting the current experience (unless switching things in `<App />` manually)

File: packages/manager/src/Router.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      globalErrors,
+      isACLPEnabled,
+      isDatabasesEnabled,
+      isPlacementGroupsEnabled,
+      selfServeBetas: flags.selfServeBetas,
+    },
+  });
+
+  return <RouterProvider router={router} />;
+};
```
Comment (abailly-akamai) [inline packages/manager/src/Router.tsx:34]: Our new entrypoint to load all routes in Cloud Manager (not active in this PR!)

File: packages/manager/src/components/AccountActivation/AccountActivationLanding.tsx [hunk 4]
```diff
@@ -69,4 +70,10 @@ const AccountActivationLanding = () => {
   );
 };
 
+export const accountActivationLandingLazyRoute = createLazyRoute(
+  '/account-activation'
+)({
+  component: AccountActivationLanding,
+});
```
Comment (abailly-akamai) [inline packages/manager/src/components/AccountActivation/AccountActivationLanding.tsx:77]: All `createLazyRoute` exports you will see in this PR aim to facilitate lazy loading smaller bundles: https://tanstack.com/router/latest/docs/framework/react/api/router/createLazyRouteFunction#createlazyroute-function

Those get imported in the /routes/* files via `.lazy()`

File: packages/manager/src/features/Billing/InvoiceDetail/InvoiceDetail.tsx [hunk 5]
```diff
@@ -157,13 +155,14 @@ export const InvoiceDetail = () => {
                 </IconButton>
               </Link>
               {invoice && (
-                <Typography
-                  data-qa-invoice-id
-                  sx={{ paddingLeft: theme.spacing(1) }}
-                  variant="h2"
-                >
-                  Invoice #{invoice.id}
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Billing/InvoiceDetail/InvoiceDetail.tsx:165]: I think the removal of the `#` is causing an e2e test failure

Comment (abailly-akamai) [reply packages/manager/src/features/Billing/InvoiceDetail/InvoiceDetail.tsx:165]: ah yes thank you, this change is unrelated but we did not have any h1 on this page so i thought to add one (was helpful during my testing to have an H1 on each page for assertions that the page is rendering something within the route. I'll fix it!

**Comment by jaalah-akamai** [comment]
@abailly-akamai Most of the routes from the top user menu give me errors, redirects, or are just not found - Are these actual issues you're seeing?

File: packages/manager/src/features/Billing/InvoiceDetail/InvoiceDetail.tsx (hunk 5 above)
Comment (abailly-akamai) [reply packages/manager/src/features/Billing/InvoiceDetail/InvoiceDetail.tsx:165]: done

File: packages/manager/src/routes/index.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import { rootRoute } from './root';
+import { searchRouteTree } from './search';
+import { stackScriptsRouteTree } from './stackscripts';
+import { supportRouteTree } from './support';
+import { volumesRouteTree } from './volumes';
+import { vpcsRouteTree } from './vpcs';
+
+const indexRoute = createRoute({
+  beforeLoad: ({ context }) => {
+    const { accountSettings } = context;
```
Comment (jaalah-akamai) [inline packages/manager/src/routes/index.tsx:31]: Nit: `accountSettings` returns as `any` - might be able to cast it using `interface AccountSettings` from our type in APIv4 or find some more elegant solution

**Review by jaalah-akamai — COMMENTED** [review]
Overall this is amazing work 🔥 - I still want to look deeper into this, but here's some minor stuff I found.

**Review by jaalah-akamai — APPROVED** [review]

**Review by bnussman-akamai — APPROVED** [review]
Overall structure and patterns look good to me!

**Pull request merged**